scylla/src/transport/locator/replication_info.rs

```diff
@@ -0,0 +1,183 @@
+use itertools::Itertools;
+
+use super::TokenRing;
+use crate::routing::Token;
+use crate::transport::node::Node;
+use std::borrow::Borrow;
+use std::collections::{BTreeSet, HashMap};
+use std::sync::Arc;
+
+#[derive(Clone)]
+pub(crate) struct DatacenterNodes {
+    pub(crate) dc_ring: TokenRing<Arc<Node>>,
+    unique_nodes_in_dc_ring: Vec<Arc<Node>>,
+    rack_count: usize,
+}
+
+#[derive(Clone)]
+pub struct ReplicationData {
+    pub(crate) global_ring: TokenRing<Arc<Node>>,
+    unique_nodes_in_global_ring: Vec<Arc<Node>>,
+
+    pub(crate) datacenters: HashMap<String, DatacenterNodes>,
+    empty_dc_data: DatacenterNodes,
+}
+
+impl ReplicationData {
+    pub fn new(ring_iter: impl Iterator<Item = (Token, Arc<Node>)>) -> ReplicationData {
+        let global_ring: TokenRing<Arc<Node>> = TokenRing::new(ring_iter);
+
+        let unique_nodes_in_global_ring = global_ring
+            .iter()
+            .map(|(_t, n)| n.clone())
+            .unique()
+            .collect();
+
+        let mut datacenter_nodes: HashMap<String, Vec<(Token, Arc<Node>)>> = HashMap::new();
+        for (token, node) in global_ring.iter() {
+            if let Some(datacenter_name) = &node.datacenter {
+                match datacenter_nodes.get_mut(datacenter_name) {
+                    Some(this_datacenter_nodes) => {
+                        this_datacenter_nodes.push((*token, node.clone()))
+                    }
+                    None => {
+                        datacenter_nodes
+                            .insert(datacenter_name.clone(), vec![(*token, node.clone())]);
+                    }
+                }
+            }
+        }
+
+        let mut datacenters: HashMap<String, DatacenterNodes> = HashMap::new();
+        for (datacenter_name, this_datacenter_nodes) in datacenter_nodes {
+            let dc_ring = TokenRing::new(this_datacenter_nodes.into_iter());
+            let unique_nodes_in_dc_ring =
+                dc_ring.iter().map(|(_t, n)| n.clone()).unique().collect();
+            // When counting racks consider None as a separate rack
+            let rack_count: usize = dc_ring
+                .iter()
+                .map(|(_t, n)| n.rack.as_ref())
+                .unique()
+                .count();
+            datacenters.insert(
+                datacenter_name,
+                DatacenterNodes {
+                    dc_ring,
+                    unique_nodes_in_dc_ring,
+                    rack_count,
+                },
+            );
+        }
+
+        let empty_dc_data = DatacenterNodes {
+            dc_ring: TokenRing::new([].into_iter()),
+            unique_nodes_in_dc_ring: Vec::new(),
+            rack_count: 0,
+        };
+
+        ReplicationData {
+            global_ring,
+            unique_nodes_in_global_ring,
+            datacenters,
+            empty_dc_data,
+        }
+    }
+
+    pub fn simple_strategy_replicas(
+        &self,
+        token: impl Borrow<Token>,
```

Comment:
IIRC I originally did `impl Borrow<Token>` because it was annoying to add dereferences/references everywhere. It doesn't hurt to accept both references and values here.